api/types.go

```diff
@@ -515,35 +530,175 @@ func (addrsS MultiaddrsSerial) ToMultiaddrs() []ma.Multiaddr {
 	return addrs
 }
 
-// Pin is an argument that carries a Cid. It may carry more things in the
-// future.
+// CidsToStrings encodes cid.Cids to strings.
+func CidsToStrings(cids []*cid.Cid) []string {
+	strs := make([]string, len(cids))
+	for i, c := range cids {
+		strs[i] = c.String()
+	}
+	return strs
+}
+
+// StringsToCidSet decodes cid.Cids from strings.
+func StringsToCidSet(strs []string) *cid.Set {
+	cids := cid.NewSet()
+	for _, str := range strs {
+		c, err := cid.Decode(str)
+		if err != nil {
+			logger.Error(str, err)
+		}
+		cids.Add(c)
+	}
+	return cids
+}
+
+// AllType is a PinType used for filtering all pin types
+const AllType PinType = -1
+
+// PinType values. See PinType documentation for further explanation.
+const (
+	// BadType type showing up anywhere indicates a bug
+	BadType PinType = iota
+	// DataType is a regular, non-sharded pin. It is pinned recursively.
+	// It has no associated reference.
+	DataType
+	// MetaType tracks the original CID of a sharded DAG. Its Reference
+	// points to the Cluster DAG CID.
+	MetaType
+	// ClusterDAGType pins carry the CID if the root node that points to
+	// all the shard-root-nodes of the shards in which a DAG has been
+	// divided. Its Reference carries the MetaType CID.
+	// ClusterDAGType pins are pinned directly everywhere.
+	ClusterDAGType
+	// ShardType pins carry the root CID of a shard, which points
+	// to individual blocks on the original DAG that the user is adding,
+	// which has been sharded.
+	// They carry a Reference to the previous shard.
+	// ShardTypes are pinned with MaxDepth=1 (root and
+	// direct children only).
+	ShardType
+)
+
+// PinType specifies which sort of Pin object we are dealing with.
+// In practice, the PinType decides how a Pin object is treated by the
+// PinTracker.
+// See descriptions above.
+// A sharded Pin would look like:
+//
+// [ Meta ] (not pinned on IPFS, only present in cluster state)
+//   |
+//   v
+// [ Cluster DAG ] (pinned everywhere in "direct")
+//   |      ..  |
+//   v          v
+// [Shard1] .. [ShardN] (allocated to peers and pinned with max-depth=1
+// | | .. |    | | .. |
+// v v .. v    v v .. v
+// [][]..[]    [][]..[] Blocks (indirectly pinned on ipfs, not tracked in cluster)
+//
+//
+type PinType int
```

Comment:
I think `PinType` and `AllPins` got mixed up from my previous comments. `PinType` should have been above the const block. And the suggestion for `AllPins` was to use the bitwise iota construction.

Reply:
ok, let me try